airo-robots/airo_robots/awaitable_action.py

```diff
@@ -0,0 +1,46 @@
+import enum
+import time
+import warnings
+from typing import Callable
+
+
+class ACTION_STATUS_ENUM(enum.Enum):
+    EXECUTING = 1
+    SUCCEEDED = 2
+    TIMEOUT = 3
+
+
+class AwaitableAction:
+    def __init__(self, done_callback: Callable[..., bool]):
+        """
+
+        Args:
+            done_callback (Callable[..., bool]): Any  callable that returns True when the action is completed.
+            In the simplest case, it can be a lambda that returns True.
+            It can also be true when the gripper is in the desired position.
+            Or it could be true after a certain amount of time has passed since the action was started.
+
+            Note that the scope of this action is to send 1 command, do some other things, then wait for the command to finish.
+            If you send multiple commands to the gripper, there is no guarantee that the gripper will execute them in the order you send them,
+             as the gripper might preempt intermediate commands after it has finished the current command.
+            Such preemption will also not be detected by this action, and hence the wait of a preempted action will either timeout or succeed by accident.
+
+        """
+        self.status = ACTION_STATUS_ENUM.EXECUTING
+        self.done_callback = done_callback
+
+    def wait(self, timeout: float = 10.0) -> ACTION_STATUS_ENUM:
+        if not self.status == ACTION_STATUS_ENUM.EXECUTING:
+            return self.status
+        while True:
+            time.sleep(0.1)
```

Comment:
Is a fixed sleep of 100 ms desirable? Maybe decrease to 1 ms?

Reply:
Agreed, was thinking about CPU usage but if you're waiting for the action, you're probably not doing something else anyways. Although 1ms intervals will be far from exact due to scheduling etc, so this could also give a false sense of 'real-time-ness'?

Reply:
I wouldn't worry to much about polling overhead at 1 poll/ms. I'd rather avoid the opposite, that when polling too slowly, the main thread might have to wait longer than necessary.

Reply:
Did some empirical measurements on paard, seems that sleeping for 1ms should be fine, going below 0.5ms is not a good idea.
![sleep_actual_vs_desired](https://user-images.githubusercontent.com/37955681/224355076-080025f0-4cd6-49a4-b10b-ad478e3f82e6.png)
![sleep_error](https://user-images.githubusercontent.com/37955681/224355097-2c84c45f-7d25-4a21-a2b7-c19911221de2.png)


It will keep the CPU more busy though, so I will probably add a parameter to the action so that you can set the desired granularity. For the gripper e.g. it does not matter, for the robot you might want to be faster.

Reply:
![sleep_error](https://user-images.githubusercontent.com/37955681/224492306-6842ad6c-09f1-4966-aa6c-80304fe6b2c2.png)
On my laptop, errros are higher though.

Reply:
On my laptop, errors are slightly higher, so I set the limit of the sleep duration to 1ms

![sleep_actual_vs_desired](https://user-images.githubusercontent.com/37955681/224494900-07440888-e565-49c8-b7ed-a81c9a5226d0.png)